xain_fl/config/schema.py

```diff
@@ -0,0 +1,358 @@
+
+       # Name of the bucket for storing the aggregated models
+       bucket = "aggregated_weights"
+
+       # AWS secret access to use to authenticate to the storage service
+       secret_access_key = "my-secret"
+
+       # AWS access key ID to use to authenticate to the storage service
+       access_key_id = "my-key-id"
+
+    This config file can be loaded and used as follow:
+
+    .. code-block:: python
+
+       from xain_fl.config import Config
+
+       config = Config.load("example_config.toml")
+
+       assert config.server.host == "[::]"
+       assert config.server.port == 50051
+
+       assert config.ai.initial_weights == "./test_array.npy"
+       assert config.ai.rounds == 1
+       assert config.ai.epochs == 1
+       assert config.ai.min_participants == 1
+       assert config.ai.fraction_participants == 1.0
+
+       assert config.storage.endpoint == "http://localhost:9000"
+       assert config.storage.bucket == "aggregated_weights"
+       assert config.storage.secret_access_key == "my-access-key"
+       assert config.storage.access_key_id == "my-key"
+    """
+
+    def __init__(self, ai: NamedTuple, storage: NamedTuple, server: NamedTuple):
+        self.ai = ai
+        self.storage = storage
+        self.server = server
+
+    @classmethod
+    def from_unchecked_dict(cls: Type[T], dictionary: Mapping[str, Any]) -> T:
+        """Check if the given dictionary contains a valid configuration, and
+        if so, create a `Config` instance from it.
+
+        Args:
+
+            dictionary: a dictionary containing the configuration
+        """
+        try:
+            valid_config = CONFIG_SCHEMA.validate(dictionary)
+        except SchemaError as err:
+            raise InvalidConfig(err)
+        return cls.from_valid_dict(valid_config)
+
+    @classmethod
+    def from_valid_dict(cls: Type[T], dictionary: Mapping[str, Any]) -> T:
+        """Create a `Config` instance for the given dictionary, assuming it
+        contains a valid configuration
+
+        Args:
+
+            dictionary: a dictionary containing the configuration
+
+        """
+        return cls(
+            AiConfig(**dictionary["ai"]),
+            StorageConfig(**dictionary["storage"]),
+            ServerConfig(**dictionary["server"]),
+        )
+
+    @classmethod
+    def load(cls: Type[T], path: str) -> T:
+        """Read the config file from the given path, check that it contains a
+        valid configuration, and return the corresponding `Config`
+        instance.
+
+        Args:
+
+            path: path to a configuration file
+        """
+        with open(path, "r") as f:
+            raw_config = toml.load(f)
+        return cls.from_unchecked_dict(raw_config)
+
+
+class InvalidConfig(ValueError):
+    """
+    Exception raised upon trying to load an invalid configuration
+    """
+
+    def __init__(self, err, *args, **kwargs):
+        super().__init__(f"Invalid configuration:\n{err.code}", *args, **kwargs)
```

Comment:
Please use standard exception message, and then structlog to log a message (our best practices).

Reply:
I'm not sure what you mean. Do you have an example somewhere else to show me?

Reply:
Also I think we don't want to print a stacktrace here. These errors are legit (as in "not an un-expected error in our code") at we should probably print a nice error message to the users instead.

Reply:
Sure, we don’t want to print here anything.
The way how we handle our exceptions (best practices in the team):
1. create a custom exception,
2. catch this exception where it can occur,
3. handle exception and also log useful information (using structlog).

Basically message is not needed in custom exception and we don’t print anything, we log errors.

Reply:
Thanks. I pushed some changes that address this I believe :)